.github/workflows/news-fragment.yml

```diff
@@ -0,0 +1,52 @@
+# Licensed to the Apache Software Foundation (ASF) under one
+# or more contributor license agreements.  See the NOTICE file
+# distributed with this work for additional information
+# regarding copyright ownership.  The ASF licenses this file
+# to you under the Apache License, Version 2.0 (the
+# "License"); you may not use this file except in compliance
+# with the License.  You may obtain a copy of the License at
+#
+#   http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing,
+# software distributed under the License is distributed on an
+# "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+# KIND, either express or implied.  See the License for the
+# specific language governing permissions and limitations
+# under the License.
+#
+---
+name: CI
+
+on:  # yamllint disable-line rule:truthy
+  pull_request:
+    types: [labeled, unlabeled, opened, reopened, synchronize]
+
+jobs:
+  check-news-fragment:
+    name: Check News Fragment
+    runs-on: ubuntu-20.04
+    if: "contains(github.event.pull_request.labels.*.name, 'airflow3.0:breaking')"
+
+    steps:
+      - uses: actions/checkout@v4
+        with:
+          persist-credentials: false
+          # `towncrier check` runs `git diff --name-only origin/main...`, which
+          # needs a non-shallow clone.
+          fetch-depth: 0
+
+      - name: Check news fragment
+        run: >
+          if ! pipx run towncrier check
+          --dir .
+          --config newsfragments/config.toml
+          --compare-with origin/${{ github.base_ref }}; then
+
+          echo "Please check
+          https://github.com/apache/airflow/blob/main/contributing-docs/16_contribution_workflow.rst
+          for guidance."
+
+          false
+
+          fi
```

Comment:
```suggestion
        run: >
          pipx run towncrier check --dir . --config newsfragments/config.toml
          --compare-with origin/${{ github.base_ref }} || 
          echo "\033[1;33mMissing significant newsfragment for PR with "airflow3.0:breaking" label. Please check 
          https://github.com/apache/airflow/blob/main/contributing-docs/16_contribution_workflow.rst 
          for guidance.\033[0m" && false"
```

Reply:
NIT: smth  like that (maybe better formatted using `run: |`. Most important is actually explaining what is missing and colored output that shows what the warning is.. Possibly  doing it Python-only in selective checks might be a good idea, it already contains changed files and PR labels as fields in SelectiveCheck object and it would avoid bash-isms and could allow using rich for color output.

Reply:
In selective checks it would be smth like - typed live so it could have errors:

``` python
if 'airflow3.0:breaking' in self._pr_labels and not filter(lambda x: x.startswith('newsfragments/significant.'), self._files):
   get_console().print("[yellow] Middinh.....")
   sys.exit(1)
```

Plud it would be easily locally testable (that's why I REALLY don't like scripts in GitHub Actions especially in `run` as they are not testable.

With breeze:

```
breeze ci selective-check --commit-ref HEAD --pr-labels '["airflow3.0:breaking"]'
```

(also unit tests could be added easily).

Reply:
`towncrier check` checks more than just the file existing; I don’t think it’s worthwhile to reimplement that just for Airflow. (It also doesn’t have a public Python API for this unfortunately.)

Reply:
Up to you - I have just seen how you struggled with it and pushed a number of changes to CI which is always pain (and bash is  something we wanted to minimise).

BTW. We could also run

```
if 'airflow3.0:breaking' in self._pr_labels:
    res = run_command("towncrier", "check", "--dir", AIRFLOW_SOURCE_DIR,  "--config", "newsfragments/config.toml",  "--compare-with HEAD^", check=False)
    if res.resultcode !=0:
        get_console()
```

Would need to add towncrier as breeze dependency, but since breeze auto-upgrades, it would be easy.

It would have the same properties for testing as explained before. When selective checks are run depth-fetch: 2 is set automatically to get list of changed files.

BTW. I **think**  - if I am not mistaken - that the bash check might have an error - it likely fails because `pipx` is not available in CI and not because towncrier check fails.

Reply:
pipx is provided by GHA images by default, that part is covered.

Reply:
I’m trying to think of a good reason to add towncrier to breeze. The check will never work locally, so requiring towncrier there feels quite wrong. It is used by the release manager though ([Airflow core](https://github.com/apache/airflow/blob/6570c6d1bb620c6a952a16743c7168c775f6ad70/dev/README_RELEASE_AIRFLOW.md?plain=1#L275) and [Helm](https://github.com/apache/airflow/blob/6570c6d1bb620c6a952a16743c7168c775f6ad70/dev/README_RELEASE_HELM_CHART.md?plain=1#L120)); would it make sense to move the release commands into `breeze release-management`?

Reply:
> pipx is provided by the ubuntu-20.04 image by default, that part is covered.

Ah yes we just upgrade it to `python -m pip install "pipx>=1.4.1` in breeze installation script to avoid the bug with 1.4.0 about setuptools(?) breaking It I think. I think at some point of time ubuntu runner had the buggy version.

> I’m trying to think of a good reason to add towncrier to breeze. The check will never work locally, so requiring towncrier there feels quite wrong. It is used by the release manager though ([Airflow core](https://github.com/apache/airflow/blob/6570c6d1bb620c6a952a16743c7168c775f6ad70/dev/README_RELEASE_AIRFLOW.md?plain=1#L275) and [Helm](https://github.com/apache/airflow/blob/6570c6d1bb620c6a952a16743c7168c775f6ad70/dev/README_RELEASE_HELM_CHART.md?plain=1#L120)); would it make sense to move the release commands into breeze release-management?

Yes. I think it makes sense  - I think having a single auto-upgradeable environment for all "release" commands has great benefits - i.e. absolute reproducibility - especially if those commands that release manager executes are also executed in CI  (which is already happenning for all release-management subcommands of breeze) - because then you avoid the surprises that things stop working at the very moment you need it as relase manager and it adds unnecessary stress - running them in CI makes a lot of sense in this case.

Reply:
I took a look in SelectiveChecks, it doesn’t really feel right to do the newsfragment check in it directly (makes debugging more difficult too). But doing the actual check in another job is a bit overcomplicating to my liking—the label check + towncrier run is quite self-contained.

I think I’m going to make the coloring changes and merge this for now. We can work on the improvements if they make sense while receiving the benefits for all PRs in the mean time.